File: frontend/projects/console-web/src/app/drafts.ts
import { Injectable, inject } from '@angular/core';
import { Observable, Subject, filter, map, startWith, switchMap } from 'rxjs';
import * as Y from 'yjs';

import { reason } from './errors';
import { Local } from './local';
import { Telemetry } from './telemetry';
import type { Picture } from './picture';

/**
 * What has been written and not sent, per conversation, shared between devices.
 *
 * A draft is a document, not a string. Two devices typing produce one text
 * holding both edits, because edits merge, here and in the runner
 * (`console/src/drafts.rs`), so nobody is asked which version they meant.
 *
 * A draft is a record, not an instruction: nothing leaves until a person presses
 * send.
 */

/** The name of the shared text inside a draft. The runner uses the same one. */
const TEXT = 'text';

/** How often to ask the runner for what it has. */
const PULL_EVERY_MS = 5000;

/**
 * How long an edit waits before it is sent.
 *
 * Not the poll interval. Leaving a keystroke to the next tick means a draft
 * can sit on one device for five seconds while the other shows nothing — and, in
 * the two-device test, it reads as the merge having failed. Short enough to feel
 * immediate, long enough that a sentence typed straight through is one push.
 */
const SEND_AFTER_MS = 300;

/** A held picture as stored: no preview, since an object URL dies with its document. */
interface HeldPicture {
  data: string;
  mediaType: string;
  width: number;
  height: number;
  bytes: number;
}

/** A stored picture, checked rather than cast: storage outlives the build that wrote it. */
function revived(value: unknown): Picture | undefined {
  if (typeof value !== 'object' || value === null) return undefined;
  const held: Partial<HeldPicture> = value;
  if (typeof held.data !== 'string' || typeof held.mediaType !== 'string') return undefined;
  if (typeof held.width !== 'number' || typeof held.height !== 'number') return undefined;
  if (typeof held.bytes !== 'number') return undefined;
  return {
    data: held.data,
    mediaType: held.mediaType,
    width: held.width,
    height: held.height,
    bytes: held.bytes,
    // A data URL rather than an object URL: nothing to revoke, and it cannot be
    // dead on arrival like a `blob:` from a previous document.
    preview: `data:${held.mediaType};base64,${held.data}`,
  };
}

/** One conversation's document and the machinery keeping it. */
interface Kept {
  readonly doc: Y.Doc;
  readonly text: Y.Text;
  /** Resolves once storage has given back what this device already had. */
  readonly loaded: Promise<unknown>;
}

function toBase64(bytes: Uint8Array): string {
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

function fromBase64(text: string): Uint8Array | undefined {
  try {
    const binary = atob(text);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
    return bytes;
  } catch {
    return undefined;
  }
}

@Injectable({ providedIn: 'root' })
export class Drafts {
  private telemetry = inject(Telemetry);
  private store = inject(Local);
  private readonly kept = new Map<string, Kept>();
  private timer?: ReturnType<typeof setInterval>;
  private since = 0;
  /** Conversations whose document has changed HERE since the last push. */
  private readonly dirty = new Set<string>();
  private sending?: Promise<void>;
  private soon?: ReturnType<typeof setTimeout>;
  /** A conversation whose picture changed, so its stream re-reads. */
  private readonly pictures = new Subject<string | null>();
  private get: typeof fetch = (...args) => fetch(...args);

  /**
   * Point this at a different runner. Called before anything else, or not at all
   * — production needs no setting up.
   */
  configure(get: typeof fetch): void {
    this.get = get;
  }

  /** The words this conversation is holding, and every change to them. */
  text$(id: string): Observable<string> {
    return new Observable<string>((to) => {
      let stop: (() => void) | undefined;
      void this.open(id).then((kept) => {
        const say = (): void => to.next(kept.text.toJSON());
        say();
        kept.doc.on('update', say);
        stop = () => kept.doc.off('update', say);
      });
      return () => stop?.();
    });
  }

  /**
   * Put `text` in the box.
   *
   * Written as a DIFF against what is there, not as a replacement. A
   * wholesale replace deletes every character and inserts the string again, which
   * merges with a concurrent edit as two people retyping the sentence at once —
   * the very mess this design exists to avoid. The common edit, a keystroke at
   * the end, becomes one insert of one character.
   */
  async write(id: string, text: string): Promise<void> {
    const kept = await this.open(id);
    const had = kept.text.toJSON();
    if (had === text) return;
    const { at, removed, added } = difference(had, text);
    kept.doc.transact(() => {
      if (removed > 0) kept.text.delete(at, removed);
      if (added) kept.text.insert(at, added);
    }, 'here');
  }

  /** Hold a picture for the next message, or put it down. */
  async hold(id: string, picture: Picture | undefined): Promise<void> {
    if (!picture) {
      await this.store.delete(`picture-${id}`);
      this.pictures.next(id);
      return;
    }
    const kept: HeldPicture = {
      data: picture.data,
      mediaType: picture.mediaType,
      width: picture.width,
      height: picture.height,
      bytes: picture.bytes,
    };
    await this.store.set(`picture-${id}`, kept);
    this.pictures.next(id);
  }

  /**
   * The picture waiting to go with the next message. Local only: a picture is
   * hundreds of kilobytes and two cannot be combined.
   */
  picture$(id: string): Observable<Picture | undefined> {
    return this.pictures.pipe(
      filter((changed) => changed === id || changed === null),
      startWith(id),
      switchMap(() => this.store.get(`picture-${id}`)),
      map(revived),
    );
  }

  /** Ask the runner for anything new, and send anything of ours it lacks. */
  sync(): void {
    this.timer ??= setInterval(() => void this.exchange(), PULL_EVERY_MS);
    void this.exchange();
  }

  /** Send what has changed here, shortly — see [SEND_AFTER_MS]. */
  private push(): void {
    if (this.soon !== undefined) return;
    this.soon = setTimeout(() => {
      this.soon = undefined;
      void this.exchange();
    }, SEND_AFTER_MS);
  }

  /** Stop syncing and let go of every document. */
  async close(): Promise<void> {
    clearInterval(this.timer);
    clearTimeout(this.soon);
    this.soon = undefined;
    this.timer = undefined;
    for (const kept of this.kept.values()) kept.doc.destroy();
    this.kept.clear();
    await Promise.resolve();
  }

  /**
   * One round: push what has changed here, then take everything the runner has
   * past our cursor.
   *
   * Both directions apply the SAME operation — merge these bytes into that
   * document — so there is no order in which this goes wrong, and no answer the
   * runner can give that has to be refused.
   */
  private async exchange(): Promise<void> {
    // One round at a time. Pushing the same edit twice would be harmless, since
    // applying an update twice is applying it once, but it is wasted bytes on a
    // phone's connection.
    this.sending ??= this.round().finally(() => {
      this.sending = undefined;
    });
    await this.sending;
  }

  private async round(): Promise<void> {
    const mine = [...this.dirty];
    this.dirty.clear();
    try {
      if (mine.length) {
        const entries = await Promise.all(
          mine.map(async (id) => ({
            ulid: id,
            update: toBase64(Y.encodeStateAsUpdate((await this.open(id)).doc)),
            at: Date.now(),
          })),
        );
        const answer = await this.get('/api/sync/drafts', {
          method: 'POST',
          headers: { 'content-type': 'application/json' },
          body: JSON.stringify(entries),
        });
        if (!answer.ok) throw new Error(`draft push failed: ${answer.status}`);
        await this.take(await answer.json());
      }
      const res = await this.get(`/api/sync/drafts?since=${this.since}`);
      if (!res.ok) throw new Error(`draft pull failed: ${res.status}`);
      const body: unknown = await res.json();
      if (typeof body !== 'object' || body === null) return;
      if ('documents' in body) await this.take(body.documents);
      // A missing or malformed cursor is LEFT ALONE rather than reset. Taking it
      // as zero would rewind the pull for ever and re-deliver the whole store every
      // five seconds.
      if ('checkpoint' in body) {
        const { checkpoint } = body;
        if (typeof checkpoint === 'object' && checkpoint !== null && 'rev' in checkpoint) {
          const { rev } = checkpoint;
          if (typeof rev === 'number') this.since = Math.max(this.since, rev);
        }
      }
    } catch (err: unknown) {
      // Said out loud: a dead tunnel is the ordinary case, and a silent failure
      // looks exactly like a quiet one from the composer. Through [[reason]], not
      // `String(err)`, which gives `[object Object]`.
      const said = reason(err);
      console.warn('draft sync:', said);
      this.telemetry.note('draft-sync', said);
      // Unsent work goes back on the pile, or an edit made during an outage would
      // never be offered again.
      for (const id of mine) this.dirty.add(id);
    }
  }

  /** Merge whatever the runner sent into the documents here. */
  private async take(documents: unknown): Promise<void> {
    if (!Array.isArray(documents)) return;
    // `unknown[]`, not the `any[]` that `isArray` narrows to: every field below
    // comes off the wire and is checked rather than trusted.
    const rows: readonly unknown[] = documents;
    for (const row of rows) {
      if (typeof row !== 'object' || row === null) continue;
      if (!('ulid' in row) || typeof row.ulid !== 'string') continue;
      if (!('update' in row) || typeof row.update !== 'string') continue;
      const update = fromBase64(row.update);
      if (!update) continue;
      const kept = await this.open(row.ulid);
      // Tagged `theirs`, which keeps it out of [dirty]: the runner's own bytes are
      // not echoed back to it.
      Y.applyUpdate(kept.doc, update, 'theirs');
    }
  }

  /** This conversation's document, restored from this device's storage. */
  private async open(id: string): Promise<Kept> {
    const already = this.kept.get(id);
    if (already) {
      await already.loaded;
      return already;
    }
    const doc = new Y.Doc();
    const kept: Kept = { doc, text: doc.getText(TEXT), loaded: this.restore(id, doc) };
    doc.on('update', (_update: Uint8Array, origin: unknown) => {
      // Anything written HERE is ours to send; anything merged in came from there.
      if (origin !== 'theirs') {
        this.dirty.add(id);
        this.push();
      }
      // Kept whole rather than appended to: a draft is a sentence, and a log of
      // updates would grow for as long as somebody kept typing without sending.
      void this.store.set(`draft-${id}`, Y.encodeStateAsUpdate(doc));
    });
    this.kept.set(id, kept);
    await kept.loaded;
    return kept;
  }

  /** Put back what this device had, if anything, before anybody reads the box. */
  private async restore(id: string, doc: Y.Doc): Promise<void> {
    const held = await this.store.get(`draft-${id}`);
    if (held instanceof Uint8Array && held.length) Y.applyUpdate(doc, held, 'theirs');
  }
}

/**
 * The one span that changed between two strings: where it starts, how much to
 * remove, and what to put there.
 *
 * The point is the COMMON case, which is typing. Adding a character at the
 * end must be one insert of one character, not a delete of the whole sentence
 * and an insert of a longer one — the second merges with a concurrent edit as
 * two people retyping at once, which is how a merging design can still lose
 * words. Anything cleverer than a shared prefix and suffix is not worth it: a
 * draft is edited by a person at a keyboard, not rewritten by a program.
 */
export function difference(
  had: string,
  wants: string,
): { at: number; removed: number; added: string } {
  let head = 0;
  // dev-lint: allow-field-identity-eq both sides are CHARACTERS of a string, not
  // fields of a record — `string[i]` is `string | undefined`, for which identity
  // IS value equality. There is no array or object a comparison could get wrong.
  while (head < had.length && head < wants.length && had[head] === wants[head]) head++;
  let tail = 0;
  while (
    tail < had.length - head &&
    tail < wants.length - head &&
    had[had.length - 1 - tail] === wants[wants.length - 1 - tail]
  ) {
    tail++;
  }
  return {
    at: head,
    removed: had.length - head - tail,
    added: wants.slice(head, wants.length - tail),
  };
}

File: console/src/drafts.rs
//! What has been written and not sent, per conversation, shared between devices.
//!
//! A draft is a record, not an instruction: it acts on nothing until a person
//! presses send, so it can be replicated freely.
//!
//! The runner holds it because both clients talk to this process. One way in,
//! `/api/sync/drafts`, in the shape life uses: two mechanisms writing one map is
//! how drafts diverge silently.
//!
//! A draft is a [`yrs`] document and a push is an update, which merges, so there
//! is no conflict to resolve. Two devices typing produce one text; the usual
//! disagreement is one device a few keystrokes behind the other.
//!
//! Text only. A picture is hundreds of kilobytes and there is no meaningful way
//! to combine two; it stays in the client's own storage.

use base64::Engine;
use parking_lot::RwLock;
use std::collections::BTreeMap;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use yrs::updates::decoder::Decode;
use yrs::{GetString, ReadTxn, Text, Transact};

/// The name of the shared text inside a draft's document. Both ends must agree on
/// it or each would read an empty string out of the other's writes.
pub const TEXT: &str = "text";

/// One conversation's unsent words, as the roster reports them.
///
/// A VIEW of the document, not the document: what the session list draws is the
/// sentence, and handing it the encoded state would put a CRDT somewhere that
/// only wants a string.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct Draft {
    /// What is being typed, read out of the merged document.
    pub text: String,
    /// Bumped on every merge, and used for ONE thing: ordering a pull. It counts
    /// across the WHOLE store, not per conversation — see [`Drafts::merge`].
    pub rev: u64,
    /// Unix milliseconds of the last merge that changed the text.
    pub at: u64,
}

/// One draft on the wire: the merged document, and the text it reads as.
///
/// `ulid` is the SESSION id — a draft is one per conversation, and the
/// conversation already has a stable identity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct DraftDoc {
    pub ulid: String,
    /// The whole document state, base64. Applying it is idempotent and
    /// order-independent, which is the whole reason this design has no conflicts: a
    /// client that applies it twice, or applies it after its own newer edit,
    /// converges either way.
    pub update: String,
    /// What [`Self::update`] reads as, for anything that wants the sentence rather
    /// than the document. Never read as input — the document is the truth.
    #[serde(default)]
    pub text: String,
    pub at: u64,
    /// Server revision, for the pull cursor. Ignored as push input; set here.
    #[serde(default)]
    pub rev: u64,
}

/// What a pull answers: the rows past the caller's checkpoint, and the new one.
#[derive(Debug, Serialize)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct PullResponse {
    pub documents: Vec<DraftDoc>,
    pub checkpoint: Checkpoint,
}

/// The pull cursor: the highest `rev` delivered so far.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct Checkpoint {
    pub rev: u64,
}

/// One change from a client: an update to merge into whatever is held.
///
/// No assumed state, and nothing to refuse. An update carries its own
/// causal context, so the runner never has to be told what the client thought was
/// here.
#[derive(Debug, Deserialize)]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export))]
pub struct PushEntry {
    pub ulid: String,
    /// The client's document state, base64.
    pub update: String,
    pub at: u64,
}

/// One conversation's document, as it is kept.
#[derive(Debug, Clone)]
struct Held {
    /// The merged state, encoded. Kept encoded rather than as a live document
    /// because that is what goes to disk and to clients, and decoding per write
    /// costs nothing at a draft's size.
    state: Vec<u8>,
    text: String,
    rev: u64,
    at: u64,
}

/// What is written to disk: [`Held`] with the state in base64, so the file stays
/// readable text like every other the console keeps.
#[derive(Debug, Serialize, Deserialize)]
struct Stored {
    update: String,
    rev: u64,
    at: u64,
}

/// Every conversation's unsent words, by session id.
#[derive(Debug)]
pub struct Drafts {
    store: PathBuf,
    held: RwLock<BTreeMap<String, Held>>,
}

/// The document an encoded update reads as, or `None` if it is not one.
///
/// A bad update is dropped, never fatal. It arrives over the wire from a
/// client that may be older than this binary, and refusing to start — or
/// poisoning the store — would turn one malformed push into an outage for every
/// conversation.
fn decoded(update: &[u8]) -> Option<yrs::Doc> {
    let doc = yrs::Doc::new();
    let update = yrs::Update::decode_v1(update).ok()?;
    doc.transact_mut().apply_update(update).ok()?;
    Some(doc)
}

/// The text inside a document.
fn reads_as(doc: &yrs::Doc) -> String {
    let text = doc.get_or_insert_text(TEXT);
    let txn = doc.transact();
    text.get_string(&txn)
}

/// A document's whole state, which is what both disk and the wire carry.
fn encoded(doc: &yrs::Doc) -> Vec<u8> {
    doc.transact()
        .encode_state_as_update_v1(&yrs::StateVector::default())
}

fn to_base64(bytes: &[u8]) -> String {
    base64::engine::general_purpose::STANDARD.encode(bytes)
}

/// Bytes out of base64, or `None` — the same tolerance as [`decoded`], for the
/// same reason.
pub fn from_base64(text: &str) -> Option<Vec<u8>> {
    base64::engine::general_purpose::STANDARD.decode(text).ok()
}

/// A document holding `text`, encoded — for a caller that has words rather than a
/// document, which is every test and nothing in production.
pub fn document_of(text: &str) -> Vec<u8> {
    let doc = yrs::Doc::new();
    let shared = doc.get_or_insert_text(TEXT);
    {
        let mut txn = doc.transact_mut();
        shared.insert(&mut txn, 0, text);
    }
    encoded(&doc)
}

impl Drafts {
    /// Read what the last run wrote. An unreadable file is an empty set and a loud
    /// line, as in [`crate::modes::Modes::load`]: the cost is a draft missing from the
    /// OTHER device, which still holds its own copy.
    pub fn load(store: PathBuf) -> Self {
        let held = match std::fs::read_to_string(&store) {
            Ok(text) => match serde_json::from_str::<BTreeMap<String, Stored>>(&text) {
                Ok(held) => held
                    .into_iter()
                    .filter_map(|(id, one)| {
                        // A row that will not decode is dropped rather than kept as an empty
                        // draft, which would look like somebody had cleared it.
                        let state = from_base64(&one.update)?;
                        let text = reads_as(&decoded(&state)?);
                        Some((
                            id,
                            Held {
                                state,
                                text,
                                rev: one.rev,
                                at: one.at,
                            },
                        ))
                    })
                    .collect(),
                Err(why) => {
                    tracing::error!(
                        "drafts: {} will not parse ({why}) — unsent words will not cross devices",
                        store.display()
                    );
                    BTreeMap::new()
                }
            },
            Err(_) => BTreeMap::new(),
        };
        Self {
            store,
            held: RwLock::new(held),
        }
    }

    /// This conversation's draft, if any device has written one.
    pub fn get(&self, id: &str) -> Option<Draft> {
        self.held.read().get(id).map(Held::seen)
    }

    /// Every draft, for the roster — a client that has just connected learns which
    /// conversations hold unsent words without asking per session.
    pub fn all(&self) -> BTreeMap<String, Draft> {
        self.held
            .read()
            .iter()
            .map(|(id, held)| (id.clone(), held.seen()))
            .collect()
    }

    /// Merge a device's document into what is held, and answer with the result.
    ///
    /// This cannot fail on a disagreement, and that is the point. Two updates
    /// written without knowledge of each other merge into one document holding both
    /// edits, and applying the same update twice changes nothing. There is no state a
    /// caller can be in that this has to refuse.
    ///
    /// `None` only when the bytes are not a document at all.
    pub fn merge(&self, id: &str, update: &[u8], at: u64) -> Option<Draft> {
        let update = yrs::Update::decode_v1(update).ok()?;
        let (one, all) = {
            let mut held = self.held.write();
            let doc = match held.get(id) {
                Some(mine) => decoded(&mine.state)?,
                None => yrs::Doc::new(),
            };
            doc.transact_mut().apply_update(update).ok()?;
            let text = reads_as(&doc);
            // One counter for the whole STORE: the pull cursor is one number across the
            // collection, and a per-document counter left a fresh conversation at rev 1
            // behind a client already at 3 — never synced, and random-looking from outside.
            let rev = held.values().map(|one| one.rev).max().unwrap_or(0) + 1;
            // A merge that changed no character does not move the clock the list dates a
            // draft by: a device can contribute history without anybody having typed.
            let at = match held.get(id) {
                Some(was) if was.text == text => was.at,
                _ => at,
            };
            let one = Held {
                state: encoded(&doc),
                text,
                rev,
                at,
            };
            held.insert(id.to_string(), one.clone());
            (one, held.clone())
        };
        self.write(&all);
        Some(one.seen())
    }

    /// Drop the drafts of conversations no longer on disk. An empty `alive` is a
    /// sweep that found nothing, not evidence that everything has gone.
    pub fn forget(&self, alive: &std::collections::BTreeSet<String>) {
        if alive.is_empty() {
            return;
        }
        let all = {
            let mut held = self.held.write();
            let before = held.len();
            held.retain(|id, _| alive.contains(id));
            if held.len() == before {
                return;
            }
            tracing::info!(
                "drafts: {} conversation(s) gone from disk, forgetting their unsent words",
                before - held.len()
            );
            held.clone()
        };
        self.write(&all);
    }

    /// Every draft past `since`, oldest revision first, with the new checkpoint. Not
    /// paged: a sentence per conversation cannot outgrow one response.
    pub fn pull(&self, since: u64) -> PullResponse {
        let held = self.held.read();
        let mut documents: Vec<DraftDoc> = held
            .iter()
            .filter(|(_, one)| one.rev > since)
            .map(|(id, one)| one.wire(id))
            .collect();
        documents.sort_by_key(|one| one.rev);
        let rev = documents.last().map_or(since, |one| one.rev);
        PullResponse {
            documents,
            checkpoint: Checkpoint { rev },
        }
    }

    /// Merge everything a client sent, and answer with the merged documents.
    ///
    /// Always the merged state, never a refusal. The client applies what comes
    /// back and is then level with the runner.
    pub fn push(&self, entries: Vec<PushEntry>) -> Vec<DraftDoc> {
        entries
            .into_iter()
            .filter_map(|entry| {
                let Some(update) = from_base64(&entry.update) else {
                    tracing::info!("{}: a draft push that is not base64", entry.ulid);
                    return None;
                };
                if self.merge(&entry.ulid, &update, entry.at).is_none() {
                    tracing::info!("{}: a draft push that is not a document", entry.ulid);
                    return None;
                }
                self.held
                    .read()
                    .get(&entry.ulid)
                    .map(|one| one.wire(&entry.ulid))
            })
            .collect()
    }

    /// Written whole each time — a short line per conversation, and a rewrite cannot
    /// leave a half-updated entry.
    fn write(&self, all: &BTreeMap<String, Held>) {
        let stored: BTreeMap<&String, Stored> = all
            .iter()
            .map(|(id, one)| {
                (
                    id,
                    Stored {
                        update: to_base64(&one.state),
                        rev: one.rev,
                        at: one.at,
                    },
                )
            })
            .collect();
        if let Ok(text) = serde_json::to_string_pretty(&stored)
            && let Some(dir) = self.store.parent()
        {
            let _ = std::fs::create_dir_all(dir);
            if let Err(why) = std::fs::write(&self.store, text) {
                tracing::warn!("drafts: could not write {}: {why}", self.store.display());
            }
        }
    }
}

impl Held {
    fn seen(&self) -> Draft {
        Draft {
            text: self.text.clone(),
            rev: self.rev,
            at: self.at,
        }
    }

    fn wire(&self, id: &str) -> DraftDoc {
        DraftDoc {
            ulid: id.to_string(),
            update: to_base64(&self.state),
            text: self.text.clone(),
            at: self.at,
            rev: self.rev,
        }
    }
}
